lib/cn/csched_sp3_work.c

```diff
@@ -196,55 +186,204 @@ sp3_work_ispill_find_kvsets(struct sp3_node *spn, uint n_max, struct kvset_list_
 static uint
 sp3_work_ispill(
     struct sp3_node *         spn,
-    uint                      cnt_min,
-    uint                      cnt_max,
+    struct sp3_thresholds    *thresh,
     struct kvset_list_entry **mark,
     enum cn_action *          action,
     enum cn_comp_rule *       rule)
 {
-    struct cn_tree_node *tn;
-    uint                 cnt;
+    struct cn_tree_node *tn = spn2tn(spn);
+    uint cnt_min, cnt_max, cnt;
+
+    if (tn->tn_loc.node_level > 0) {
+        cnt_min = thresh->ispill_kvsets_min;
+        cnt_max = thresh->ispill_kvsets_max;
+    } else {
+        cnt_min = thresh->rspill_kvsets_min;
+        cnt_max = thresh->rspill_kvsets_max;
+    }
 
-    cnt = sp3_work_ispill_find_kvsets(spn, cnt_max * 4, mark);
+    cnt = sp3_work_ispill_find_kvsets(spn, cnt_max, mark);
     if (cnt < cnt_min)
         return 0;
 
     *action = CN_ACTION_SPILL;
-
-    tn = spn2tn(spn);
+    *rule = CN_CR_ISPILL;
 
     if (tn->tn_loc.node_level > 0) {
+        uint compc = kvset_get_compc((*mark)->le_kvset);
+        uint64_t clen = cn_ns_clen(&tn->tn_ns);
+
+        /* If there is another job currently spilling this node then wait
+         * to schedule additional jobs until there are a sufficient number
+         * of kvsets to spill.
+         */
+        if (compc < 2 && cnt < cnt_max / 2) {
+            if (atomic_read(&tn->tn_busycnt) > 0)
+                return 0;
+        }
+
+        /* If the spill size is large and the oldest kvset is heavily
+         * compacted then spill only the oldest kvset.  If the oldest
+         * kvset is lightly compacted then just reduce the spill count.
+         *
+         * Otherwise, if the spill size is tiny then defer the spill
+         * until there are a sufficient number of kvsets to spill.
+         */
+        if (clen > tn->tn_size_max / 2 ||
+            cn_ns_keys(&tn->tn_ns) > ((uint64_t)thresh->ispill_pop_keys << 20)) {
+
+            if (compc > 2) {
+                *rule = CN_CR_ISPILL_ONE;
+                cnt = 1;
```

Comment:
breaking spills into smaller operations results in more k-compaction and write amp in the child nodes.

Reply:
This usually only occurs when a node is first being spilled and thus heavily compacted, so shouldn't be an issue.  On the flip side, we now no longer spill kvsets one-by-one after the level spill, so that should actually result in less space and write amp :-)   This is all now tunable via ispill_pop_szgb and ispill_pop_keys rparams, so easily to tweak if we don't like the results.